File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        pair = {
            ')':'(',
            '}':'{',
            ']':'['
        }

        for c in s:
            if c not in pair: # if c is open bracket
                stack.append(c) 
            # elif c is close bracket
            elif not stack or pair[c] != stack.pop():
                return False
                
        return True
        
    

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    

    const map = new Map([
        [')','('],
        [']','['],
        ['}','{']
    ]);
    let stack:Array<string> = [];

    for(const char of s) {
        if(map.has(char)) {
            if(stack.length === 0 || stack.pop() !== map.get(char)) {
                return false;
            }
        } else {
            stack.push(char);
        }
    }
    
    return stack.length === 0;
};